### PR TITLE
Allow setting activation time in contest.yaml

### PR DESCRIPTION
**Pull request by vmcj**
Closes https://github.com/DOMjudge/domjudge/issues/1719

I've added some more times to the contest which might be useful. If we can expose a property but currently it's null, should we display as null or do we omit it from the                                response? I've now chosen to omit as that is also easier to parse for downstream clients. It's easy to crash an a missing key and fix it, but finding why you get a strange time when converting null to something else is slightly harder IMO.

### CODE REVIEW

File: webapp/src/Service/ImportExportService.php [hunk 1]
```diff
@@ -148,9 +148,20 @@ public function importContestData($data, ?string &$message = null, string &$cid
 
         $starttime->setTimezone(new DateTimeZone(date_default_timezone_get()));
         $activateTime = new DateTime();
-        if ($activateTime > $starttime) {
+        if (array_key_exists('activate-time', $data)) {
+            $activatetimeValue = $data['activate-time'];
+            if (is_string($activatetimeValue)) {
```
Comment (nickygerritsen) [inline webapp/src/Service/ImportExportService.php:153]: We have this logic somewhere else as well right? Maybe put it in Utils?

File: webapp/src/Service/ImportExportService.php [hunk 2]
```diff
@@ -148,9 +148,20 @@ public function importContestData($data, ?string &$message = null, string &$cid
...
+                    // Make sure ISO 8601 but with the T replaced with a space also works.
+                    date_create_from_format('Y-m-d H:i:sO', $activatetimeValue);
+            } else {
+                /** @var DateTime $activateTime */
+                $activateTime = $activatetimeValue;
+            }
+        } elseif ($activateTime > $starttime) {
             $activateTime = $starttime;
         }
+        $activateTime->setTimezone(new DateTimeZone(date_default_timezone_get()));
```
Comment (eldering) [inline webapp/src/Service/ImportExportService.php:164]: Why this change? This is also changing the behavior of the old case, is that intentional?

File: webapp/src/Service/ImportExportService.php (hunk 1 above)
Comment (vmcj) [reply webapp/src/Service/ImportExportService.php]: I've rewrote the PR, so I'll resolve this.

File: webapp/src/Service/ImportExportService.php (hunk 2 above)
Comment (vmcj) [reply webapp/src/Service/ImportExportService.php]: I've rewrote the PR, so I'll resolve this.

**Review by nickygerritsen — APPROVED** [review]

File: webapp/src/Controller/API/ContestController.php [hunk 3]
```diff
@@ -417,13 +417,25 @@ public function getContestYamlAction(Request $request, string $cid): StreamedRes
         $response->setCallback(function () use ($contest, $penalty_time) {
             echo "name:                     " . $contest->getName() . "\n";
             echo "short-name:               " . $contest->getExternalid() . "\n";
+            echo "activation-time:               " .
```
Comment (eldering) [inline webapp/src/Controller/API/ContestController.php:420]: Alignment is off here and below.

File: webapp/src/Controller/API/ContestController.php [hunk 4]
```diff
@@ -417,13 +417,25 @@ public function getContestYamlAction(Request $request, string $cid): StreamedRes
...
+                Utils::absTime($contest->getActivatetime(), true) . "\n";
             echo "start-time:               " .
                 Utils::absTime($contest->getStarttime(), true) . "\n";
             echo "duration:                 " .
                 Utils::relTime($contest->getEndtime() - $contest->getStarttime(), true) . "\n";
             echo "scoreboard-freeze-length: " .
```
Comment (eldering) [inline webapp/src/Controller/API/ContestController.php:426]: Slightly unrelated, but we should change this to `scoreboard_freeze_duration`, see https://ccs-specs.icpc.io/draft/contest_api#contests and https://ccs-specs.icpc.io/draft/contest_package#example-yaml-files

File: webapp/src/Controller/API/ContestController.php [hunk 5]
```diff
@@ -417,13 +417,25 @@ public function getContestYamlAction(Request $request, string $cid): StreamedRes
         $response->setCallback(function () use ($contest, $penalty_time) {
             echo "name:                     " . $contest->getName() . "\n";
             echo "short-name:               " . $contest->getExternalid() . "\n";
```
Comment (eldering) [inline webapp/src/Controller/API/ContestController.php:419]: Here and below: we should use `_` (underscore) instead of `-` (dash).

File: webapp/src/Controller/API/ContestController.php [hunk 6]
```diff
@@ -417,13 +417,25 @@ public function getContestYamlAction(Request $request, string $cid): StreamedRes
...
             echo "scoreboard-freeze-length: " .
                 Utils::relTime($contest->getEndtime() - $contest->getFreezetime(), true) . "\n";
             echo "penalty-time:             " . $penalty_time . "\n";
+            echo "freeze-time:               " .
+                Utils::absTime($contest->getFreezetime(), true) . "\n";
+            echo "end-time:               " .
+                Utils::absTime($contest->getEndtime(), true) . "\n";
+            echo "finalize-time:               " .
+                Utils::absTime($contest->getFinalizetime(), true) . "\n";
+            echo "unfreeze-time:               " .
```
Comment (eldering) [inline webapp/src/Controller/API/ContestController.php:435]: We may want to call this `scoreboard_thaw_time` to already prepare for https://github.com/icpc/ccs-specs/pull/134/files#diff-353f5220dc395fec383c89d1431e3c5733913a1af93f9e8d3c0135d49b2fb5ebR735

**Comment by vmcj** [comment]
I had to break the import of the old import format for this, I assume we want to support that again as we also still support the TSV files?

File: webapp/src/Service/ImportExportService.php [hunk 7]
```diff
@@ -83,15 +99,47 @@ public function getContestYamlData(Contest $contest, bool $includeProblems = tru
         return $data;
     }
 
+    protected function convertImportedTime(array $fields, array $data, ?string &$message = null): ?DateTimeImmutable
+    {
+        $timeValue = null;
+        $usedField = null;
+        foreach ($fields as $field) {
+            if ($timeValue) {
+                continue;
```
Comment (eldering) [inline webapp/src/Service/ImportExportService.php:108]: Why not just break as soon as `$timeValue` gets set?

File: webapp/src/Service/ImportExportService.php [hunk 8]
```diff
@@ -83,15 +99,47 @@ public function getContestYamlData(Contest $contest, bool $includeProblems = tru
         return $data;
     }
 
+    protected function convertImportedTime(array $fields, array $data, ?string &$message = null): ?DateTimeImmutable
```
Comment (eldering) [inline webapp/src/Service/ImportExportService.php:102]: Would be good to briefly comment what this does: find the first set field from `$fields` in `$data` and parse it as a date.
Also, would be good to clarify that there's a distinction betwee a return value of `null` and parse failure, and that that's only the case when `$message` is not `null`. Would be better to rename that to `$error_message` or so.

File: webapp/src/Service/ImportExportService.php [hunk 9]
```diff
@@ -83,15 +99,47 @@ public function getContestYamlData(Contest $contest, bool $includeProblems = tru
...
     public function importContestData($data, ?string &$message = null, string &$cid = null): bool
     {
         if (empty($data) || !is_array($data)) {
             $message = 'Error parsing YAML file.';
             return false;
         }
 
-        $requiredFields = [['start_time', 'start-time'], 'name', ['id', 'short-name'], 'duration'];
-        $missingFields  = [];
+        $activateTimeFields = ['activate-time', 'activation-time', 'activate_time', 'activation_time'];
```
Comment (eldering) [inline webapp/src/Service/ImportExportService.php:138]: I think these should be ordered from most recent standard to oldest/most non-standard. So `activate_time` first.

File: webapp/src/Service/ImportExportService.php [hunk 10]
```diff
@@ -83,15 +99,47 @@ public function getContestYamlData(Contest $contest, bool $includeProblems = tru
...
+        $deactivateTimeFields = preg_filter('/^/', 'de', $activateTimeFields);
+        $startTimeFields = ['start_time', 'start-time'];
+        $requiredFields = [$startTimeFields, 'formal_name', ['id', 'name', 'short-name', 'short_name'], 'duration'];
```
Comment (eldering) [inline webapp/src/Service/ImportExportService.php:141]: `formal_name` should not be required, we should just set our `name` to either `name` from the `contest.yaml` or if that doesn't exist the `formal_name`. So I guess it should be `['name', 'formal_name']` in there.

File: webapp/src/Service/ImportExportService.php [hunk 11]
```diff
@@ -117,39 +165,43 @@ public function importContestData($data, ?string &$message = null, string &$cid
...
-            $activateTime = $starttime;
+        $deactivateTime = $this->convertImportedTime($deactivateTimeFields, $data, $message);
+        if ($message) {
+            return false;
         }
+
         $contest = new Contest();
         $contest
-            ->setName($data['name'])
+            ->setName($data['formal_name'])
```
Comment (eldering) [inline webapp/src/Service/ImportExportService.php:191]: This should use `name` and fall back to `formal_name`.

File: webapp/src/Service/ImportExportService.php (hunk 7 above)
Comment (vmcj) [reply webapp/src/Service/ImportExportService.php:108]: Original author didn't do that, but makes sense IMO

**Comment by mpsijm** [comment]
Thanks for merging this! :smile:

Question: since either `activat{e,ion}{_,-}time` is allowed as key, what is the canonical form? I would guess `activation_time`, but I also see `deactivate_time` used in [the code](https://github.com/DOMjudge/domjudge/pull/1807/files#diff-2e92e4f26fbadf9eff221ac3fa36ef1bf0bf1077b805645896e361aa5db841c9R78), so asking just to be sure :slightly_smiling_face:

**Comment by eldering** [comment]
It's `(de)activate_time`, see https://github.com/DOMjudge/domjudge/blob/main/webapp/src/Entity/Contest.php#L84

For most other properties you can refer to https://ccs-specs.icpc.io/draft/contest_api#interface-specification, but these are our own specification.
